app/web_framework/cluster_slave_application.py

```diff
@@ -116,6 +126,82 @@ def get(self, build_id, subjob_id):
         self.write(response)
 
 
+class _AtomsHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id):
+        response = {
+            'comment': 'not implemented',
+            }
+        self.write(response)
+
+
+class _AtomHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id, atom_id):
+        response = {
+            'comment': 'not implemented',
+            }
+        self.write(response)
+
+
+class _AtomConsoleHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id, atom_id, max_lines=50, offset_line=None):
+        """
+        Return the console output if it exists, returns a 404 otherwise.
+
+        On success, the response contains keys: offset_line, num_lines, total_num_lines, and content.
+
+        e.g.:
+        {
+            'offset_line': 0,
+            'num_lines': 50,
+            'total_num_lines': 167,
+            'content': 'Lorem ipsum dolor sit amet,\nconsectetur adipiscing elit,\n...',
+        }
+
+        :type build_id: int
+        :type subjob_id: int
+        :type atom_id: int
+        :param max_lines: The maximum total number of lines to return. If this max_lines + offset_line lines do not
+            exist in the output file, just return what there is.
+        :type max_lines: int
+        :param offset_line: The line number (0-indexed) to start reading content for. If none is specified, we will
+            return the console output starting from the end of the file.
+        :type offset_line: int | None
+        """
+        if offset_line is not None and offset_line < 0:
+            response = {'message': '\'offset_line\' must be greater than or equal to zero.'}
+            self._write_status(response, success=False, status_code=400)
+            return
+        if max_lines <= 0:
+            response = {'message': '\'max_lines\' must be greater than zero.'}
+            self._write_status(response, success=False, status_code=400)
+            return
+
+        artifact_dir = BuildArtifact.atom_artifact_directory(build_id, subjob_id, atom_id)
+        output_file = os.path.join(artifact_dir, BuildArtifact.OUTPUT_FILE)
+
+        # If the console output file doesn't exist where we expect it to be, return a 404.
+        if not os.path.isfile(output_file):
+            response = {'message': 'Output file doesn\'t exist for build_id: {} subjob_id: {} atom_id: {}'.format(
+                build_id, subjob_id, atom_id)}
+            self._write_status(response, success=False, status_code=404)
```

Comment:
Continuing from my comment above, a 404 can be set by raising an `ItemNotFoundError`. The full list is at the top of the `ClusterBaseHandler` class.